riverdriver/riverpgxv5/internal/dbsqlc/river_job.sql

```diff
@@ -0,0 +1,315 @@
+        finalized_at = CASE WHEN state = 'running'::river_job_state THEN finalized_at ELSE now() END,
+        -- Mark the job as cancelled by query so that the rescuer knows not to
+        -- rescue it, even if it gets stuck in the running state:
+        metadata = jsonb_set(metadata, '{cancel_attempted_at}'::text[], @cancel_attempted_at::jsonb, true)
+    FROM notification
+    WHERE river_job.id = notification.id
+    RETURNING river_job.*
+)
+SELECT *
+FROM river_job
+WHERE id = @id::bigint
+    AND id NOT IN (SELECT id FROM updated_job)
+UNION
+SELECT *
+FROM updated_job;
+
+-- name: JobDeleteBefore :one
+WITH deleted_jobs AS (
+    DELETE FROM river_job
+    WHERE id IN (
+        SELECT id
+        FROM river_job
+        WHERE
+            (state = 'cancelled' AND finalized_at < @cancelled_finalized_at_horizon::timestamptz) OR
+            (state = 'completed' AND finalized_at < @completed_finalized_at_horizon::timestamptz) OR
+            (state = 'discarded' AND finalized_at < @discarded_finalized_at_horizon::timestamptz)
+        ORDER BY id
+        LIMIT @max::bigint
+    )
+    RETURNING *
+)
+SELECT count(*)
+FROM deleted_jobs;
+
+-- name: JobGetAvailable :many
+WITH locked_jobs AS (
+    SELECT
+        *
+    FROM
+        river_job
+    WHERE
+        state = 'available'::river_job_state
+        AND queue = @queue::text
+        AND scheduled_at <= now()
+    ORDER BY
+        priority ASC,
+        scheduled_at ASC,
+        id ASC
+    LIMIT @max::integer
+    FOR UPDATE
+    SKIP LOCKED
+)
+UPDATE
+    river_job
+SET
+    state = 'running'::river_job_state,
+    attempt = river_job.attempt + 1,
+    attempted_at = now(),
+    attempted_by = array_append(river_job.attempted_by, @attempted_by::text)
+FROM
+    locked_jobs
+WHERE
+    river_job.id = locked_jobs.id
+RETURNING
+    river_job.*;
+
+-- name: JobGetByKindMany :many
+SELECT *
+FROM river_job
+WHERE kind = any(@kind::text[])
+ORDER BY id;
+
+-- name: JobGetByKindAndUniqueProperties :one
+SELECT *
+FROM river_job
+WHERE kind = @kind
+    AND CASE WHEN @by_args::boolean THEN args = @args ELSE true END
+    AND CASE WHEN @by_created_at::boolean THEN tstzrange(@created_at_start::timestamptz, @created_at_end::timestamptz, '[)') @> created_at ELSE true END
+    AND CASE WHEN @by_queue::boolean THEN queue = @queue ELSE true END
+    AND CASE WHEN @by_state::boolean THEN state::text = any(@state::text[]) ELSE true END;
+
+-- name: JobGetByID :one
+SELECT *
+FROM river_job
+WHERE id = @id
+LIMIT 1;
+
+-- name: JobGetByIDMany :many
+SELECT *
+FROM river_job
+WHERE id = any(@id::bigint[])
+ORDER BY id;
+
+-- name: JobGetStuck :many
+SELECT *
+FROM river_job
+WHERE state = 'running'::river_job_state
+    AND attempted_at < @stuck_horizon::timestamptz
+LIMIT @max;
+
+-- name: JobInsert :one
+INSERT INTO river_job(
+    args,
+    attempt,
+    attempted_at,
+    created_at,
+    errors,
+    finalized_at,
```

Comment:
Per earlier conversation, I think there's a good argument for minimizing these two insert queries so that they only have fields which are intended to be set during initial insert. We can save the fields like `errors`, `finalized_at`, `attempt`, and `attempted_at` for a separate insert function only used by internal tests.

This would keep this queries shorter and eliminate some unnecessary data being sent over the wire on every insert. Then we can add a separate insert query that lets you set _any_ field (for testing) but which isn't actually used by any application code.

Reply:
+1.

Just finished breaking everything up into three functions:

* `JobInsertFast`
* `JobInsertFastMany`
* `JobInsertFull`

The reason I'd added the extra insert parameters to `JobInsertMany` was to keep functionality parity with `JobInsert`. Now that the "full" path is separate, `JobInsertFastMany` can use the same minimal parameter set as `JobInsertFast`. `JobInsertFull` is just for tests and doesn't need a "many" variation (I guess we could conceivably add one for tests that want to add multiple jobs, but that's definitely another yak to shave).

A bonus from the refactor is that even compared to `master`, I was able to strip three more parameters from `JobInsertFast` (`Attempt`, `AttemptedAt`, `CreatedAt`). These were used in tests, but not in the non-test code.

Reply:
> (I guess we could conceivably add one for tests that want to add multiple jobs, but that's definitely another yak to shave).

Seems unnecessary to do this for just a minimal test perf boost, I doubt we'll ever find it worthwhile.

Reply:
Yep, agreed.